Support for additional file integrity checksum algorithms for cloud services

mrpasquini opened the pull request:
### Motivation / Background

S3, GCS, and Azure all support file integrity checksums other than MD5. The intent of this PR is to support these to ActiveStorage service classes and centralize hashing functionality.

### Detail

- Goal was to improve FIPS support #34443 by adding additional methods of file integrity hashing and remove #54040
- Make ActiveStorage service adapters responsible for checksums and implementation of hashing
- Add ActiveStorage::Checksum class to keep digest/algorithm coupled
- Replace ActiveStorage::Blob checksum with backwards compatible encoded ActiveStorage::Checksum
- Expand direct upload interface to allow passing of algorithm + digest as either ActiveStorage::Checksum compatible string format `MD5:vckNNU4TN7zbzl+o3tjXPQ==` or hash `checksum: { algorithm: "MD5", digest: "vckNNU4TN7zbzl+o3tjXPQ=="}`. Keep backwards compatibility of existing string format as MD5 digest

### Additional information

- Supports all algorithms for Azure, GCS, & S3 ~except for CRC64-NVMe for S3. [PR merged](https://github.com/postmodern/digest-crc/pull/45) to Digest::CRC 7.0~

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

brosintoski commented on activestorage/lib/active_storage/service/azure_storage_service.rb:
```diff
@@ -26,14 +31,15 @@ def initialize(storage_account_name:, storage_access_key:, container:, public: f
       @signer = Azure::Storage::Common::Core::Auth::SharedAccessSignature.new(storage_account_name, storage_access_key)
       @container = container
       @public = public
+      @checksum_algorithm = checksum_algorithm
```
should this not default to MD5 to avoid regressions?

byroot commented on activestorage/activestorage.gemspec:
```diff
@@ -38,4 +38,5 @@ Gem::Specification.new do |s|
   s.add_dependency "activerecord",  version
 
   s.add_dependency "marcel",    "~> 1.0"
+  s.add_dependency "digest-crc"
```
It would be best to avoid a new dependency that wouldn't be used by most people.

In general what we do is try to require, rescue `LoadError` etc.

mrpasquini replied:
I think this can be removed as its already a dependency of the cloud service SDK gems..

Earlopain replied:
That is also not a dependency of rails. This only passes in tests since it's part of the lockfile but a fresh app with ActiveStorage will error on the eager requires.

mrpasquini replied:
Is it better to do as singleton on ActiveStorage like MD5 per type? https://github.com/rails/rails/blob/9f8d180f7014b83b3eb8d3cf93fc093725b712b9/activestorage/lib/active_storage.rb#L368-L375

or is it better to memoize in ActiveStorage::Checksum class like
```
class ActiveStorage::Checksum
  require "openssl"

  attr_accessor :crc32
...
    def for(checksum_algorithm)
      case checksum_algorithm
...
      when :CRC32
        return @crc32 if @crc32
        begin
          require "digest/crc32"
        rescue LoadError
          raise LoadError, 'digest/crc32 not loaded. Please add `gem "digest-crc"` to your gemfile.'
        end
        @crc32 = Digest::CRC32
...
```

byroot replied:
The thing is, for services that support something better than MD5, we should just use that better thing and not rely on the global property.

e.g. google-cloud-storage can use `crc` and depends on the `digest-crc` gems, we should consider just always using that, I don't see any reason to keep using MD5 for those.

Azure it a bit more tricky because it doesn't depend on the digest crc gem.

In other words, as long as we can stop using MD5 without requiring data migration, we should do it.

mrpasquini replied:
I was initially not planning to change defaults in this PR, but it makes sense if we handle cases where the stored `checksum` in the `active_storage_blobs` table doesnt match the current config for the service in this PR
Azure recommends CRC64
GCS recommends CRC32C
S3 recommends CRC64-NVMe. I submitted a [PR](https://github.com/postmodern/digest-crc/pull/45) digest/crc for add there.

skipkayhil replied:
The Azure backend is deprecated and will be removed before 8.1 is released anyways, so I don't think its worth changing/thinking too hard about.

byroot commented on activestorage/app/models/active_storage/blob.rb:
```diff
@@ -331,8 +331,9 @@ def service
+      return unless service.try(:checksum_algorithm)
 
-      ActiveStorage.checksum_implementation.new.tap do |checksum|
+      ActiveStorage::Checksum.for(service.try(:checksum_algorithm)).new.tap do |checksum|
```
I'd rather avoid `try` here, it's a big smell.

mrpasquini replied:
Removed in favor of checking if `service` is nil

byroot commented on activestorage/app/models/active_storage/checksum.rb:
```diff
@@ -0,0 +1,26 @@
+  require "digest/crc32c"
+  require "digest/crc64"
+
+  SUPPORTED_CHECKSUMS = {
+    MD5: ActiveStorage.md5_checksum_implementation,
+    SHA1: OpenSSL::Digest::SHA1,
+    SHA256: OpenSSL::Digest::SHA256,
+    CRC32: Digest::CRC32,
+    CRC32c: Digest::CRC32c,
+    CRC64: Digest::CRC64
+  }
```
This way of doing things impose early loading of `CRC*` even if they're never used.

byroot commented on activestorage/lib/active_storage/service/azure_storage_service.rb:
```diff
@@ -197,5 +203,16 @@ def handle_errors
+        when :MD5
+          { "Content-MD5" => checksum }
+        when :CRC64
+          { "x-ms-content-crc64" => checksum }
```
I'm not very familiar with Active Storage. This `checksum` is only used for upload? Not stored in DB or anything? 

In otherwords, can we switch to other algorithms, or does it require a complex data migration for users?

mrpasquini replied:
It is 
- used by the cloud service to validate the uploaded data or return an error.
- saved locally as part of `active_storage_blobs` table in `checksum` column. I can add another column to store the type to support migrations/transitioning from one type to another.

byroot replied:
> saved locally as part of active_storage_blobs table in checksum column.

Ah that's what I feared.

Rather than another column we could do a prefix, e.g. `md5:caffe42343...`, and if no prefix we can assume it's MD5 which would allow for a smooth transition hopefully? (again I apologize here, I never used Active Storage beyond basic feature so I'm not as familiar with its internal than with other Rails parts).

Earlopain replied:
There have been a few migrations for AS in the past (https://github.com/rails/rails/tree/main/activestorage/db/update_migrate). This one in particular is very close what would be needed here https://github.com/rails/rails/blob/fa9cf269191c5077de1abdd1e3f934fbeaf2a5d0/activestorage/db/update_migrate/20190112182829_add_service_name_to_active_storage_blobs.rb

It seems fine to add a new column for this? I guess the question would be if that column is public. In my application, I grab that checksum and decode it to display the actual md5 hash in the UI which would break when it contains additional data. Mind you, it would also break when it contains a different hash type, but less so and it would be more egonomic to handle if there were two dedicated columns for this.

mrpasquini replied:
I also played around with a branch that added a new column for the algorithm attribute, but I prefer this approach with serialize due to backwards compatibility being more simple.

byroot commented:
@mrpasquini this is a bit much for me to review given my limited familiarity with Active Storage, and how big this change is.

But I'll try to give some comments.

byroot commented on activestorage/app/models/active_storage/checksum.rb:
```diff
@@ -0,0 +1,128 @@
+# frozen_string_literal: true
+
+class ActiveStorage::Checksum # :nodoc:
```
Perhaps it would make sense to implement `to_s` and `to_str` to return `@digest`?

Or perhaps it's better to fail if `checksum` is used as a string?

mrpasquini replied:
I only saw the one test fail in ActionText and that looked like a test expectation and not something that would have actual functionality. Is it more of an issue if we return `@digest` and it is not an MD5 digest as expected?

byroot replied:
That's why I'm a bit uncomfortable making such a change in Active Storage, I haven't used it enough in production myself to have a good idea of what users do with `blob.checksum`.

I suspect it's not directly access by the vast majority of users, but I don't have a strong opinion.

mrpasquini replied:
I can see the backwards compatibility case for MD5 digest with to_s, but I am not sure I have a strong on to_s return values beyond that.

 I may lean toward using `dump`.  We would match the value stored in the checksum database column and would be accepted by the direct upload action param.
```
class ActiveStorage::Checksum
...
def to_s
  self.class.dump(self)
end
```

or we could raise an error if not MD5... but this seems more disruptive.
```
class ActiveStorage::Checksum
...
def to_s
  digest if algorithm == :MD5
  raise StandardError # or some other type?
end
```

byroot commented on activestorage/lib/active_storage/service/azure_storage_service.rb:
```diff
@@ -12,9 +12,14 @@ module ActiveStorage
   # Wraps the Microsoft Azure Storage Blob Service as an Active Storage service.
   # See ActiveStorage::Service for the generic API documentation that applies to all services.
   class Service::AzureStorageService < Service
```
Might not be worth bothering with `azure` as IIRC we plan to extract it.

byroot commented on activestorage/app/models/active_storage/checksum.rb:
```diff
@@ -0,0 +1,128 @@
+  end
+
+  class << self
+    attr_accessor :crc32_class, :crc32c_class, :crc64_class, :md5_class
```
These accessors seem unused?

mrpasquini replied:
Yes, these can now be removed. They are being using these to memoize the below methods so we dont have to repeatedly make the require call and are not accessed directly.

byroot commented on activestorage/app/models/active_storage/checksum.rb:
```diff
@@ -0,0 +1,128 @@
+    def dump(checksum)
+      "#{checksum.algorithm}:#{checksum.digest}" if checksum
+    end
+
```
```suggestion
```

byroot commented on activestorage/app/models/active_storage/blob.rb:
```diff
@@ -21,6 +21,7 @@ class ActiveStorage::Blob < ActiveStorage::Record
 
   has_secure_token :key, length: MINIMUM_TOKEN_LENGTH
   store :metadata, accessors: [ :analyzed, :identified, :composed ], coder: ActiveRecord::Coders::JSON
+  serialize :checksum, coder: ActiveStorage::Checksum
```
One issue with this is that after upgrading all existing blob will be marked as `changed?`. In general it's not a huge deal, but it can cause some nasty corner cases.

It could be interesting to ensure `dump/load` perfectly rountrip old values.

mrpasquini replied:
If it is important to preserve the raw MD5 value in the DB moving forward I can add a case to `dump` to write out in that format.

byroot commented:
I talked a bit with @zzak, and he suggested something I find interesting.

Rather than make this much change in Active Storage itself, each adapter could chose to use whatever it wants. Unless I'm mistaken nothing in the Active Storage API says `checksum` must be MD5, if so we can probably refactor it out.

What I like about this idea, is we could start experimenting with a single adapter, making it a far smaller and easier PR.

mrpasquini commented on activestorage/app/models/active_storage/blob.rb:
```diff
@@ -325,23 +326,11 @@ def purge_later
 
   # Returns an instance of service, which can be configured globally or per attachment
   def service
+    return unless service_name
```
Existing test will fail
https://github.com/rails/rails/blob/main/activestorage/test/models/blob_test.rb#L305-L311

mrpasquini commented on activestorage/app/models/active_storage/variant.rb:
```diff
@@ -71,7 +71,7 @@ def processed
   # Returns a combination key of the blob and the variation that together identifies a specific variant.
   def key
-    "variants/#{blob.key}/#{OpenSSL::Digest::SHA256.hexdigest(variation.key)}"
+    "variants/#{blob.key}/#{ActiveStorage::Checksum.sha256.hexdigest(variation.key)}"
```
I can remove this and the SHA1 in variation. I was trying to provide a stable, centralized place to monkey patched in alternate/acceptable implementations as needed for a FIPS environment, but I believe we have until Dec. 31, 2030 until SHA1 is deprecated, so there is likely time to deal with that in the future.

mrpasquini commented on activestorage/app/models/active_storage/checksum.rb:
```diff
@@ -0,0 +1,147 @@
+      send(checksum_algorithm.downcase) if SUPPORTED_CHECKSUMS.include?(checksum_algorithm)
+    end
+
+    def file(file, algorithm)
```
Could you share more reasoning to use kwarg here? If algorithm wasnt provided would we then raise an ArguementError?

mrpasquini commented on activestorage/test/controllers/direct_uploads_controller_test.rb:
```diff
@@ -15,7 +17,7 @@ class ActiveStorage::S3DirectUploadsControllerTest < ActionDispatch::Integration
 
     test "creating new direct upload" do
-      checksum = ActiveStorage.checksum_implementation.base64digest("Hello")
+      checksum = ActiveStorage::Checksum.base64digest("Hello", ActiveStorage::Blob.service.checksum_algorithm)
```
Let me try adding a `file` and `base64digest` to the service classes.

mrpasquini commented:
> I talked a bit with @zzak, and he suggested something I find interesting.
> 
> Rather than make this much change in Active Storage itself, each adapter could chose to use whatever it wants. Unless I'm mistaken nothing in the Active Storage API says `checksum` must be MD5, if so we can probably refactor it out.
> 
> What I like about this idea, is we could start experimenting with a single adapter, making it a far smaller and easier PR.

I am happy to break this PR into multiple. It has grown rather large. I see some of the refactoring and exploration contributing to that.
- MD5 assumptions have been pushed to the edges at data parsing as means to maintain backwards compatibility at both the stored data level (no schema or data migrations) and with the direct upload interface (no breaking change to React or other front end apps).
- Since algorithm/digest are together in the ActiveStorage::Checksum instance, there are no assumptions or differences in behavior based on algorithm except in the service adapters.
- Existing blobs can be verified/processed and new blobs can be created with supported algorithms that are not the service's default algorithm. This means existing blobs would continue to function if you want to move from MD5 to CRC64NVMe or a legacy web app using direct upload could continue to use MD5 even if you moved the default to SHA256.

To make sure I understand the other item you mention, you are wanting the service/adapters to load and manage the classes for algorithms? So S3Service would support and manage loading for `:CRC32, :CRC32c, :MD5, :SHA1, :SHA256, :CRC64NVMe` rather than a dedicated class like ActiveStorage::Checksum found in this PR? Would that result in a lot of duplicate code across the services (which maybe isnt a big deal as there are only a few and it seems like you are trying to extract them out of AS). What would we do for something like the disk class where I would expect it to support the superset of available digests from the other adapters? Would it have to be an explicitly maintained and growing list?

mrpasquini committed:
Make ActiveStorage::Service responsible for checksums

mrpasquini committed:
Refactor MD5 fallback for ActiveStorage to service

mrpasquini committed:
Add ActiveService::Checksum and make Service responsible for hashing

mrpasquini committed:
Use Checksum class with Adapters

mrpasquini committed:
Update CHANGELOG.md

mrpasquini committed:
Add file check algorithms for S3

- add CRC32, CRC32c, SHA1, SHA256, and CRC64NVMe to S3 service
- Add default_digest_algorithm to S3 config
- remove now unneeded load/to_s wrappers on S3 service test checksums.
- S3 tests pass respecting default_digest_algorithm setting

mrpasquini committed:
Add CRC32c to GCS service

mrpasquini committed:
Add CRC64 to Azure service

mrpasquini committed:
Remove compatibility shims

mrpasquini committed:
Add file check algorithms for DiskService

Add CRC32, CRC32c, SHA1, SHA256, CRC64, CRC64NVMe support to DiskService

byroot commented:
> So S3Service would support and manage loading for `:CRC32, :CRC32c, :MD5, :SHA1, :SHA256, :CRC64NVMe` rather than a dedicated class like ActiveStorage::Checksum found in this PR?

Yes.

> Would that result in a lot of duplicate code across the services

A bit, but presumably each service class would only check for one or two algorithm.

> What would we do for something like the disk class

The disk service should probably be opiniated and select an algorithm that comes builtin in the `digest` gem, and isn't too likely to cause problems in the future. Probably `SHA256`?

mrpasquini commented:
I had moved responsibility and implementation of hashing to the services earlier this week. I left MD5 in the base service class so everything supports it for backwards compatibility.

The ActiveStorage::Checksum class is just the coder and digest/algorithm class replacing Blob.checksum's string attribute.

I reorganized the commits so I can split this PR after the first:
[Make ActiveStorage::Service responsible for checksums](https://github.com/rails/rails/pull/54123/commits/08e959147f9a4a6befc84862bcf98872242f35a4)

after the 3rd and/or 4th
[Add ActiveService::Checksum and make Service responsible for hashing](https://github.com/rails/rails/pull/54123/commits/d7d5bc4a6d05741aba1c0c603840c89c5d934a92)
[d7d5bc4](https://github.com/rails/rails/pull/54123/commits/d7d5bc4a6d05741aba1c0c603840c89c5d934a92)
[Use Checksum class with Adapters](https://github.com/rails/rails/pull/54123/commits/024c26049f237f4d7bfc9dde3ddd5af0237d65eb)

and then for each of the services.

mrpasquini commented:
Closing this PR as it has gotten rather large. Opened #54468 as first digestible unit of work toward this feature.
